File: backend/utils/template_helpers.py
"""
Helper functions for working with app templates
"""
from typing import Dict, Any


def flatten_object(obj: Dict[str, Any], prefix: str = '') -> Dict[str, Any]:
    """
    Flattens a nested object structure into a flat object with dot notation keys
    Example: { a: { b: 1 } } becomes { 'a.b': 1 }
    
    Args:
        obj: The object to flatten
        prefix: The prefix to use for keys (used in recursion)
        
    Returns:
        A flattened object
    """
    result = {}
    for key, value in obj.items():
        pre = f"{prefix}.{key}" if prefix else key
        if isinstance(value, dict):
            result.update(flatten_object(value, pre))
        else:
            result[pre] = value
    return result


def replace_placeholders(template: str, config: Dict[str, Any]) -> str:
    """
    Replaces placeholders in a template string with values from a configuration object
    Handles both flat and nested objects by flattening the configuration first
    
    Args:
        template: The template string with {{placeholder}} syntax
        config: The configuration object with values to replace placeholders
        
    Returns:
        The template with placeholders replaced
    """
    # First handle nested objects by flattening them
    flat_config = flatten_object(config)
    
    # Then replace all placeholders in the template
    result = template
    
    # First replace nested object references (e.g., {{app_settings.app_title}})
    for key, value in flat_config.items():
        placeholder = f"{{{{{key}}}}}"
        result = result.replace(placeholder, str(value))
    
    # Then replace any remaining direct references (e.g., {{app_title}})
    # This is for backward compatibility with templates that don't use nested references
    for key, value in config.items():
        if not isinstance(value, dict):
            placeholder = f"{{{{{key}}}}}"
            result = result.replace(placeholder, str(value))
    
    return result

File: frontend/admin-dashboard/src/utils/templateHelpers.ts
/**
 * Helper functions for working with app templates
 */

/**
 * Flattens a nested object structure into a flat object with dot notation keys
 * Example: { a: { b: 1 } } becomes { 'a.b': 1 }
 * 
 * @param obj The object to flatten
 * @param prefix The prefix to use for keys (used in recursion)
 * @returns A flattened object
 */
export function flattenObject(obj: Record<string, any>, prefix = ''): Record<string, any> {
  return Object.keys(obj).reduce((acc: Record<string, any>, k: string) => {
    const pre = prefix.length ? `${prefix}.` : '';
    if (typeof obj[k] === 'object' && obj[k] !== null && !Array.isArray(obj[k])) {
      Object.assign(acc, flattenObject(obj[k], pre + k));
    } else {
      acc[pre + k] = obj[k];
    }
    return acc;
  }, {});
}

/**
 * Replaces placeholders in a template string with values from a configuration object
 * Handles both flat and nested objects by flattening the configuration first
 * 
 * @param template The template string with {{placeholder}} syntax
 * @param config The configuration object with values to replace placeholders
 * @returns The template with placeholders replaced
 */
export function replacePlaceholders(template: string, config: Record<string, any>): string {
  // First handle nested objects by flattening them
  const flatConfig = flattenObject(config);
  
  // Then replace all placeholders in the template
  let result = template;
  
  // First replace nested object references (e.g., {{app_settings.app_title}})
  Object.entries(flatConfig).forEach(([key, value]) => {
    const placeholder = new RegExp(`\\{\\{${key}\\}\\}`, 'g');
    result = result.replace(placeholder, String(value));
  });
  
  // Then replace any remaining direct references (e.g., {{app_title}})
  // This is for backward compatibility with templates that don't use nested references
  Object.entries(config).forEach(([key, value]) => {
    if (typeof value !== 'object' || value === null) {
      const placeholder = new RegExp(`\\{\\{${key}\\}\\}`, 'g');
      result = result.replace(placeholder, String(value));
    }
  });
  
  return result;
}
